### ISSUE TITLE
[feature] Provide `conanrun.env` file when using `VirtualRunEnv` generator.

### ISSUE THREAD

**Issue by hwhsu1231** [conan-io/conan#11042]
I use VSCode with [CMake-Tools](https://github.com/microsoft/vscode-cmake-tools) plugin to write/debug the CMake project. After the executable is generated, it needs to find those **DLL** files on Windows when using `xxx:shared=True` option. If we don't want to use **[imports](https://docs.conan.io/en/latest/reference/conanfile_txt.html#imports)** section to copy those DLL file to where executable locates, we can use `VirtualRunEnv` generator to use `PATH` to find them. According to the reply: https://github.com/conan-io/conan/issues/11040#issuecomment-1101357775

> Yes, you could use the VirtualRunEnv generator, to generate scripts that will add those paths to the PATH and after sourcing those generated scripts the system will find the dll's for those libraries, please check the [VirtualRunEnv documentation](https://docs.conan.io/en/latest/reference/conanfile/tools/env/virtualrunenv.html) for more detailed information.

However, although `VirtualRunEnv` generators provide scripts(`conanrun.(bat|sh)`) to setup the environment variables, it seems that VSCode doesn't provide the features to preload those `*.bat` or `*.sh` file in `launch.json`, dynamically. The only way to do that is to use `envFile` option to load `*.env` file to setup environment.

If `VirtualRunEnv` generator can generate `conanrun.env` additionally, that would be helpful when using `launch.json` to debug the executable. For example:

![image](https://user-images.githubusercontent.com/71438617/163899736-b94b43a6-8616-463c-9474-5c2d32117973.png)

**Comment by hwhsu1231** [comment]
I manually copy the `PATH` environment setup from `conanrunenv-debug-x86_64.bat` to `conanrun.env`, and it works **"successfully"** without using **imports** section to copy those DLL files.

![image](https://user-images.githubusercontent.com/71438617/163902707-d99372a1-1d4c-4720-80b0-344d283fdb2b.png)
![image](https://user-images.githubusercontent.com/71438617/163902606-55715a0a-281e-4a3f-b9be-37eb080af4be.png)
![image](https://user-images.githubusercontent.com/71438617/163902718-aa5eb40c-132d-4218-a9c0-88e87a12fccb.png)

**Comment by hwhsu1231** [comment]
I found this issue, which might be related to mine:

- https://github.com/conan-io/conan/issues/8714

I'm confused that **"Does Conan already have other features to generate `.env` file?"**

**Comment by memsharded** [comment]
Yes, we have been prioritizing so far the terminal usage and the core Conan integration, being correct over the environment activation. Integrations with the IDEs is still to be checked for the new environment, and it makes sense to have some feature that generates .env files (maybe besides the .sh/.bat ones). I have annotated it for discussion shortly after 2.0. In the meantime, I think it is relatively easy to generate the file directly with the ``generate()`` method, and instantiating the ``VirtualRunEnv()``, then iterating its ``.vars`` to dump to a file.

**Comment by hwhsu1231** [comment]
@memsharded May I ask, you add this issue to the 2.X milestone, will this feature be added in Conan 1.X, recently?

**Comment by SpaceIm** [comment]
It's also something I've suggested here: https://github.com/conan-io/conan/issues/10075#issuecomment-978064194

Until conan can provide some IDE integration, I use this `tasks.json`: https://github.com/SpaceIm/vscode-cpp-cmake-conan-template/blob/master/.vscode/tasks.json

Basically, I use this task to run conan install with many generators, parse the conanrun script file generated by conan, generate a dotenv file, and give it to `launch.json` (do not run executables through `CMake: Launch`).

Full instructions: https://github.com/SpaceIm/vscode-cpp-cmake-conan-template#readme

**Comment by maximMalofeev** [comment]
It would be great if the `EnvVars` contained a method like `save_env` similar to the existing ones like `save_bat`, `save_ps1` and `save_sh`, which do the following: 
```
result = []
for varname, varvalues in self._values.items():
    result.append("{}={}".format(varname, varvalues))

content = "\n".join(result)
open(file_location, "w", encoding="utf-8").write(content)
```

This will allow to simply make .env file in the `generate()` method like 
```
re = VirtualRunEnv(self)
re.generate()
re.save_env(re.basename + ".env")
```

**Comment by memsharded** [comment]
Hi @maximMalofeev 

This ticket is kind of an old one. 
Nowadays, the modern ``CMakeToolchain`` is able to generate CMakePresets.json files with env-vars defined, so that would be useful for a large majority of cases. What is exactly your use case? The same as suggested above by the OP?

**Comment by maximMalofeev** [comment]
Yes, the case is the same as above. I need a way to provide environment variables for the vscode debug target in the `launch.json`.

**Comment by jjcasmar** [comment]
@memsharded Althought having CMakePresets with the correct environment set is a nice addition, there are still some cases where it is not enough. 
I am writing right now a pybind11 module and I need to create a vscode launch.json to launch /usr/bin/python which runs the script containing my C++ module. Since my module depends on other C++ libraries, I need some way to say where to find them.

**Comment by JonasToth** [comment]
@memsharded The need for this ticket is still there and maybe you could point me the right direction for a PR. Given that the `CMakeToolchain` generator internally holds the proper environment variables, isn't there a way to let it generate the `.env` file I manually generated via `save_env` and an additional generator? I couldn't find the proper spot in the conan code (and my immediate problem was solved so I did not invest further).

---

I stumbled across this ticket during porting of a large code base to linux, conan2 and vscode/nvim/whatever setup using LSP and DAP.
Thanks to @maximMalofeev for the inspiration :)

In the projects `conanfile.py` I added the following pieces:

```py
from conan.tools.env import VirtualRunEnv

def save_env(virtualRunEnv):
    result = []
    for varname, varvalues in virtualRunEnv.vars().items():
        result.append("{}={}".format(varname, varvalues))

    content = "\n".join(result)
    open(virtualRunEnv.basename + ".env", "w", encoding="utf-8").write(content)

class externalLibraries(ConanFile):
    build_policy = "missing"
    settings = "os", "compiler", "build_type", "arch"
    # ...
    def generate(self):
        tc = CMakeToolchain(self, generator="Ninja")
        # ...
        tc.generate()

        # NEW for `conanrunenv.env`
        env = VirtualRunEnv(self)
        save_env(env)
        env.generate()
```

This allows the following workflow:

```bash
$ conan install .
> In this project it creates:
> .bin/gcc/Debug/generators/conanrunenv.env

$ cmake --preset lin-gcc-debug
$ # ...
```

And the `.vscode/launch.json` contains the following pieces:
```json
{
    "configurations": [
        {
            "type": "lldb",
            "request": "launch",
            "name": "Debug",
            "envFile": "${workspaceFolder}/.bin/gcc/Debug/generators/conanrunenv.env",
            "initCommands": ["command source '${workspaceFolder}/tools/lldb/visualizers.lldb'"],
            "program": "${command:cmake.launchTargetPath}",
            "args": [],
            "stopOnEntry": false,
            "cwd": "${workspaceFolder}"
        }
    ]
}
```

This allows a seemless debugging experience with in a `CMake` + `conan2` + `VSCode` / `nvim` setup.

**Comment by memsharded** [comment]
Hi @JonasToth 

Thanks for the feedback.
Yes, we were recently reconsidering the creation of ``.env`` files, and we agreed that it was good to give it a try and propose the addition.
I am trying to look for the ticket, but I can't find it now, will have another look.

**Comment by memsharded** [comment]
I haven't managed to find that ticket/conversation :(

In any case, @JonasToth if you want to give it a try, these would be some guidelines:

- The generation can be implemented in the ``EnvVars`` class, where other ``save_bat()``, ``save_sh()`` files are located, and those methods could serve as a guide
- dotenv files syntax allows interpolation (using existing environment values), that should be used
- I would probably make the generation of the ``.env`` files an opt-in, via a conf like ``tools.env.virtualenv:dotenv=True``
- All contributions need unit tests to cover the functionality. "Integration" tests would be the recommended for this feature, in the ``test/integration/environment/test_dotenv.py`` new file. You can check other files, but the ``TestClient()`` helper should help a lot in writing cases, we can help in this if necessary.

**Comment by JonasToth** [comment]
@memsharded If would like it even more, if the `CMakeToolchain` generator gets that option to generate `.env` files. Does it follow the same structure as you explained or do I need to retrieve the variables somehow?

I don't want to promise that I work on. If anyone is quick at the job or motivated, please go ahead. But I think I will give it a shot at a calm hour and try to figure out how to do this in conan :)

**Comment by memsharded** [comment]
> @memsharded If would like it even more, if the CMakeToolchain generator gets that option to generate .env files. Does it follow the same structure as you explained or do I need to retrieve the variables somehow?

The ``.env`` feature is something that cannot be tied specifically to CMake build system, as it will also be useful for other build systems, and even for cases where no build-system is involved, like capturing the runtime for an application and launching it or some tests for it that require a run virtualenv, and the launcher is expecting ``.env`` files.

It can be leveraged by the generated CMakePresets.json if it makes sense, but that is a second derivative of the feature.

> I don't want to promise that I work on. If anyone is quick at the job or motivated, please go ahead. But I think I will give it a shot at a calm hour and try to figure out how to do this in conan :)

Sure, no commitment at all, feel free. I will leave it assigned to you and me at the moment (to guarantee that no one else from the team would take it, very unlikely, but just in case), if you see you can't do it, just let us know and we will try to do it ourselves when priorities allow.

**Comment by memsharded** [comment]
Hi all

I have done an initial implementation for this feature, please see https://github.com/conan-io/conan/pull/18266, feedback very welcome!

**Comment by samuel-emrys** [comment]
Hi @memsharded,

Just to add to this use case - I have the same use case as described in the OP. 

## Aims

I'm trying to set up a reasonably coherent conan/cmake workspace with VSCode, with the following aims:

1. Be able to execute CMake Configure stage
2. Be able to execute CMake Build stage
3. Be able to execute CMake Test stage
4. Be able to specify a build/execution target and run that target
5. Be able to specify breakpoints and debug a running application and step into dependencies

## Issues

Of the above aims, (1) and (2) are resolved without awareness of the run environment, and all build environment information is contained within the CMakePresets.json.

However, the presence of the `run` environment is key to (3), (4) and (5) if there are any shared libraries in the tree, and (5) when there's a desire to step into the source code of a dependency whilst debugging.

Fortunately, the `run` environment is specified in the test preset, and so my requirement for this feature doesn't include (3), since tests are appropriately aware of the `run` environment.

However, the way that the VSCode Extension [CMake Tools](https://github.com/microsoft/vscode-cmake-tools/tree/main) deals with (4) and (5) doesn't appear to be through the CMakePresets context - instead it suggests [setting up a `launch.json`](https://github.com/microsoft/vscode-cmake-tools/blob/015abf52297c5319fb3e81b598f4b11dd82479b0/docs/debug-launch.md#debug-using-a-launchjson-file) file to describe how the application should be launched and which debugger to attach.

## Current workaround

The current workflow that I have to resolve this is to:

```
$ source build/Release/generators/conanrun.sh
$ code .
```

Which is a bit clunky because it requires closing `code` and going through the same process every time I want to generate a different configuration.

## Solution

Fortunately, it seems like there's a reasonably easy solution in just providing the `launch.json` file with a `conanrun.env` file, since `launch.json` supports specification of dotenv files [through the `envFile` field](https://code.visualstudio.com/docs/cpp/launch-json-reference#_environment-variable-definitions-file). This is the solution that @SpaceIm [proposed above](https://github.com/SpaceIm/vscode-cpp-cmake-conan-template/blob/master/.vscode/launch.json#L11):

```json
...
            "name": "C++ GNU Debug",
            "type": "cppdbg",
            "request": "launch",
            "program": "${command:cmake.launchTargetPath}",
            "args": [],
            "cwd": "${command:cmake.launchTargetDirectory}",
            "envFile": "${command:cmake.buildDirectory}/.conan/conanrun.env",
            "MIMode": "gdb",
            "osx": {
                "MIMode": "lldb"
            }
...
```

However, this solution relies on a [short one liner script](https://github.com/SpaceIm/vscode-cpp-cmake-conan-template/blob/master/.vscode/tasks.json#L7) to convert a `conanrun.bat` or a `conanrun.sh` into a `conanrun.env`. This works great for an individual, but scaling out this becomes problematic - the use of powershell as the default on windows makes an assumption about the default terminal used in VSCode (a lot of our team use bash on Windows, too).

In this case, having conan generate the `conanrun.env` file seems like the best solution. I could modify the `global.conf` that I keep in our config repo to include `tools.env.virtualenv:dotenv=True` and provide a seamless conan/cmake workflow with VSCode.

## Some of the other possible solutions that I've considered:

### 1. Modifying the `generate()` function for each package to write the `conanrun.env` file

Unfortunately this isn't really sustainable. I manage in the order of 200-300 conan packages. Replicating that code in every recipe will be a maintenance nightmare.

### 2. Writing a custom generator

This is more feasible, since I can write it once and maintain a common definition in our config repo. I don't think this is the best solution though because it's invasive. Without modifying the recipes, this would require every user to append `-g dotenv` to their initial `conan install` command, lest they lose their ability to execute and debug the program. It's workable, but more error prone, especially when it will need to be done pretty much _every time_.

I haven't yet tested the change proposed in #18266, but having reviewed the tests that you wrote it looks like it will satisfy my requirement. It's a 👍 from me. I would start using this feature as soon as it becomes available as it eases an acute point of friction with the development flow (restarting VSCode).

**Comment by memsharded** [comment]
Thanks very much for the comprehensive feedback @samuel-emrys 


Then, I will be trying to move forward the ``dotenv`` file generation in https://github.com/conan-io/conan/pull/18266, but if you could check or run some proof of concept that it actually works for your flow, then that would help too. Otherwise, I might try to release it anyway, and hopefully it doesn't need to be changed or fixed breaking too much.

**Comment by samuel-emrys** [comment]
I just wanted to provide some quick feedback after having tested this briefly:
1. The `.env` files are created successfully 🥳 
2. The files are named `conan{run,build}env-debug-x86-64.env` - not a deal breaker, but it would be nice to have a name that doesn't encode the build type. For single configuration builds this is already in the path name. This means that I would need to rely on additional variable injection to select the right file, i.e.

```diff
-            "envFile": "${command:cmake.buildDirectory}/generators/conanrun.env"
+            "envFile": "${command:cmake.buildDirectory}/generators/conanrunenv-${command:cmake.buildType}-x86-64.env"
```
The challenge here is casing - the file uses lowercase `debug` rather than `Debug`, which is the `buildType` - this will likely cause issues on linux where it's case sensitive. I think it would be better to have a variant that either loads the fully encoded file with a similar name (same pattern as with `sh`/`bat`), or just name them `conanrun.env` and `conanbuild.env`.
3. VSCode fails to read the file, with the following error:
```
Invalid variable reference ${PATH} in string: ${CONAN_DOTENV_FOLDER}\\..\\..\\..\\.conan\\p\\b\\zlib1a4c8caac273b\\p\\bin;${PATH}.
Invalid variable reference ${CONAN_DOTENV_FOLDER} in string: ${CONAN_DOTENV_FOLDER}\\..\\..\\..\\.conan\\p\\b\\zlib1a4c8caac273b\\p\\bin;${PATH}.
```

The contents of the generated file is:

```
CONAN_DOTENV_FOLDER="B:\username\projects\conan_vS_debug_test\build\Debug\generators"
PATH="${CONAN_DOTENV_FOLDER}\..\..\..\.conan\p\b\zlib1a4c8caac273b\p\bin;${PATH}"
```
I'm testing this on a simple project; pretty much the same as `https://github.com/conan-io/examples2.git/tutorial/consuming_packages/simple_cmake_project` with `zlib` as the only dependency.

I'm still not sure why this is the case - I've done some quick research and it's turned up nothing - it seems like it should be valid. I'll continue searching.

Edit: removing the variable substitution and replacing them with the strings that they evaluate to works, so it does appear to be a problem that visual studio has in variable substitution within `envFile` - at least with the current syntax.

**Comment by memsharded** [comment]
> The files are named conan{run,build}env-debug-x86-64.env - not a deal breaker, but it would be nice to have a name that doesn't encode the build type. For single configuration builds this is already in the path name. This means that I would need to rely on additional variable injection to select the right file, i.e.

The filenames might be simplified and even remove the ``build_type`` and ``arch`` part.
But I was hoping that for this use case, where there is some tool support like IDEs managing the environment from dotenv files, that it would be possible to manage different environments per configuration, because sometimes this is the case. 
While this wouldn't be an issue for single-config projects like Unix Makefiles, multi-config projects like Visual Studio ones will have just one "build/generators" folder, without any distinction in the folder names, so they either use a different filename or they will overwrite each other and will only allow 1 configuration for the environment.


> Edit: removing the variable substitution and replacing them with the strings that they evaluate to works, so it does appear to be a problem that visual studio has in variable substitution within envFile - at least with the current syntax.

Uhm, interesting discovery.
It might be that VS variable substitution doesn't happen line by line, it would be useful to know if variable substitution is happening at all, with some pre-existing system env-var, or if it is just happening for variable substitution of variables defined in the current file.
In theory from https://hexdocs.pm/dotenvy/dotenv-file-format.html#values, it seems that this is valid syntax,
```
SIMPLE=xyz123
INTERPOLATED="Multiple\nLines and variable substitution: ${SIMPLE}"
```
maybe a bug in the environment processing in VS?

**Comment by samuel-emrys** [comment]
> maybe a bug in the environment processing in VS?

Yes possibly. I've raised https://github.com/microsoft/vscode/issues/257570 to see if variable substitution/interpolation is something that's actually supported by VSCode. The more I've looked into this, it's not clear to me that VSCode's `envFile` support is _actually_ a dotenv file - it seems more likely to me that this is a proprietary format that loosely aligns with the dotenv spec. Their [official docs](https://code.visualstudio.com/docs/cpp/launch-json-reference#_environment-variable-definitions-file) are very light on, and indicate that they don't support multiline variable specification (which is supported by the dotenv spec).

Having said that, it would make sense that they would deviate slightly since they want to support their own variable syntax as well. I found that this syntax _does_ work:

```
PATH="B:\username\projects\conan_vS_debug_test\build\Debug\generators\..\..\..\.conan\p\b\zlib1a4c8caac273b\p\bin;${env:PATH}"
```

Note the `${env:PATH}`, which suggests that they do support variable substitution with their own variable syntax. It's still not clear how we might reference other variables defined within the `.env` file though.

In terms of conan support, there's a pathway forward (in lieu of a response from microsoft) by adopting some combination of the following:

1. Use VSCode's variable syntax where required
2. Expand all variables so that there's no cross referencing/interpolation within `conanrun.env`

I think that neither are ideal. Especially (1), which would change the scope of this to being a general `dotenv` generator to a `vscode.env` generator or similar, which might reduce the usage. (2) however, whilst not ideal, would still allow the `.env` file to be generally useful with only some minor consideration for the fact that any interpolated values derived from the user environment (i.e. `PATH`) would only be up to date as at generation time.

I would find (2) to be an acceptable compromise.

> While this wouldn't be an issue for single-config projects like Unix Makefiles, multi-config projects like Visual Studio ones will have just one "build/generators" folder

Good point, I hadn't considered that - I've mentally written multi-config generators off in my head after running into too many edge cases from a developer workflow perspective/inconsistencies in workflow between linux and windows.

**Comment by memsharded** [comment]
The VSCode https://github.com/microsoft/vscode/issues/257570 ticket was closed in favor of https://github.com/microsoft/vscode-cpptools/issues/13781, which it seems it has been acknowledged as a bug, but at the moment apparently not prioritized to be solved.

I have done a few updates:

- Made the name of the generated file to be cased as the build type: ``conanbuildenv-Release.env`` to make it easier to do substitution
- I have removed the variable placeholder for interpolation, replaced it with the actual value of the env-var. This might have the issues mentioned above, but probably good to start moving it
- Adding a warning about ``Experimental`` status and about the interpolation:
  ```
   WARN: experimental: Creating dotenv file: conanbuildenv-Release.env
   Files generated with absolute paths, not interpolated.
   When https://github.com/microsoft/vscode-cpptools/issues/13781 solved, it will get interpolation
   ````


I think with these changes it might be good to go, but it would be good to have your feedback: @samuel-emrys @jjcasmar @maximMalofeev @JonasToth @hwhsu1231 @SpaceIm

**Comment by memsharded** [comment]
This has been closed by https://github.com/conan-io/conan/pull/18266, which will be available in Conan 2.21, with the approach in the previous comment. Feedback very welcome, please open new tickets for any further question or issue.

**Comment by Kaaserne** [comment]
For future reference, I'd like to add that 
```diff
- "envFile": "${command:cmake.buildDirectory}/generators/conanrunenv-${command:cmake.buildType}-x86-64.env"
+ "envFile": "${command:cmake.buildDirectory}/generators/conanrunenv-${command:cmake.buildType}-x86_64.env"
```
`-x86-64.env` -> `-x86_64.env`. Gave me an error in VSCode because it couldn't find `-x86-64.env`.

**Comment by memsharded** [comment]
Hi @Kaaserne 

If this is an issue, please open a new tickets with a bit more of details to understand and reproduce it. Many thanks!

**Comment by Kaaserne** [comment]
Not really an issue but more of a comment to a typo that was made above: x64-86 should've been x64_86

**Comment by memsharded** [comment]
Well, depending on where it is that line, it would be a "bug" or not. For example a typo in the docs that affects code examples or command line examples are "bugs". Where is that line?

 >"envFile": "${command:cmake.buildDirectory}/generators/conanrunenv-${command:cmake.buildType}-x86-64.env"

Where did you find it?

**Comment by Kaaserne** [comment]
Over here (this thread) https://github.com/conan-io/conan/issues/11042#issuecomment-3105696577

**Comment by memsharded** [comment]
Ah, ok, then only in this thread, not in Conan code or docs. Great, thanks!
